proposals/NNNN-retroactive-protocol-refinement.md

```diff
@@ -0,0 +1,100 @@
+  }
+}
+```
+In practical terms, for the Swift compiler, this effectively means creating ad-hoc extensions on Int8, Int16 and all types which conform to FixedWidthInteger along with their associated witnesses. It's more powerful however as an idea and in more abstract terms (from the Generics Manifesto):
+
+```
+protocol P {
+  func foo()
+}
+
+protocol Q {
+  func bar()
+}
+
+extension Q : P { // Make every type that conforms to Q also conforms to P
+  func foo() {    // Implement `P.foo` requirement in terms of `Q.bar`
+    bar()
+  }
+}
+
+func f<T: P>(t: T) { ... }
+
+struct X : Q {
+  func bar() { ... }
+}
+
+f(X()) // okay: X conforms to P through the conformance of Q to P
+```
+```
+This is an extremely powerful feature: it allows one
+to map the abstractions of one domain into another
+domain (e.g., every Matrix is a Graph).
+```
+It continues though:
+
+```
+However, similar to private conformances, it puts a
+major burden on the dynamic-casting runtime to chase
+down arbitrarily long and potentially cyclic chains
+of conformances, which makes efficient implementation
+nearly impossible.
+```
+The PR above strives to demonstrate that it may be possible to avoid these pitfalls and refine protocols retroactively at compile time, providing summary conformance data to the existing runtime for an efficient implementation of dynamic casts.
+
+## Proposed solution
+
+Allow conformances to be added to protocol extensions with all that implies in terms of all nominals conforming to the protocol being extended acquiring the added conformances. It would be useful if this also worked for synthesized conformances (and indeed it does).
+
+## Detailed design
+
+The PR above makes changes in three principal areas internal to the compiler.
+
+1)  The ConformanceLookupTable — a cache of the protocols a nominal conforms to has been adapted to traverse the inherited conformances of protocol extensions.
+
+2) The "Generic Signature" builder and module deserializing code has been adapted to traverse these extended sources of conformances.
+
+3) While the above takes place, a registry of extended conformances is kept in the compiler. This is used to emit the ad-hoc witness tables used by extended conformances in a manner that is compatible across modules.
+
+In terms of what is surfaced to the language, extending conformances of a protocol is fully functional including across modules.
+
+## Source compatibility
+
+This is an additive feature that has a syntax which while it follows the norms of extensions in the Swift language it is currently not permitted. Therefore, there is no effect on source compatibility.
+
+## Effect on ABI stability
+
+This is a compiler level addition to the language that does not require any changes to the Swift runtime in order to work so it has no effect on ABI stability.
+
+## Effect on API resilience
+
+This proposal does not make additions to the public API but is rather is a change to the capabilities of the compiler.
```

Comment:
IIUC, this section is meant to be a discussion of whether/how library authors may use the proposed feature in a resilient library. Specific to this proposal, I think the **Effect on API resilience** section should communicate whether library authors may add or remove retroactive refinements in an API compatible (and ABI compatible, for libraries built in evolution mode) way.

Reply:
Thanks for the comment, the proof of concept and resilient protocols is not a good mix indeed you can’t really use the feature to extend the conformances of a protocol in the standard library which is resilient as it needs to know the witness table layout to extend. Ultimately this precludes a lot of use cases even though they may work for now. Can you think of a wording that would capture this?

Reply:
Is this limitation just a by-product of the current prototype implementation, or is it an inherent limitation of the feature as proposed? If it's the latter, I would say something like,

> Retroactive refinement is not available for protocols declared in libraries with evolution mode enabled. If such a retroactive conformance is attempted, the compiler will emit an error of the form "<appropriate error message here>".

I think you should also address the compatibility implications of adding a new retroactive refinement *within* a resilient library. For instance, the current [Library Evolution page](https://swift.org/blog/library-evolution/#examples-of-resilient-changes) has this to say about protocols (among other items):

> - New protocol conformances can be added to classes, structs, and enums. (Even if they’re @frozen.)
> - Changes to a protocol’s list of refined protocols are not allowed.

So this section of your proposal should answer the question, "can a resilient library add a retroactive conformance to an existing protocol?". This is subtly different than the question that your comment addresses, which is, "can a client of a resilient library retroactively refine a protocol from that library?". Both questions should be addressed!

Reply:
I’ve updated the proposal to outline this limitation in more detail.